cabal-install/src/Distribution/Client/ProjectConfig/Parsec.hs

```diff
@@ -0,0 +1,418 @@
+
+                  rest <- go [] xs
+                  pure . fmap mconcat . sequence $ [fs, res', rest]
+          )
+          (parseImport pos importLines)
+      (Section (Name _pos name) args xs') | name == "if" -> do
+        subpcs <- go [] xs'
+        let fs = fmap singletonProjectConfigSkeleton $ fieldsToConfig source (reverse acc)
+        (elseClauses, rest) <- parseElseClauses xs
+        let condNode =
+              (\c pcs e -> CondNode mempty mempty [CondBranch c pcs e])
+                <$> parseConditionConfVar args
+                <*> subpcs
+                <*> elseClauses
+        pure . fmap mconcat . sequence $ [fs, condNode, rest]
+      _ -> go (x : acc) xs
+    go acc [] = do
+      normSource <- canonicalizeConfigPath projectDir source
+      pure . fmap singletonProjectConfigSkeleton . fieldsToConfig normSource $ reverse acc
+
+    parseElseClauses :: [Field Position] -> IO (ParseResult (Maybe ProjectConfigSkeleton), ParseResult ProjectConfigSkeleton)
+    parseElseClauses x = case x of
+      (Section (Name _pos name) _args xs' : xs) | name == "else" -> do
+        subpcs <- go [] xs'
+        rest <- go [] xs
+        pure (Just <$> subpcs, rest)
+      (Section (Name _pos name) args xs' : xs) | name == "elif" -> do
+        subpcs <- go [] xs'
+        (elseClauses, rest) <- parseElseClauses xs
+        let condNode =
+              (\c pcs e -> CondNode mempty mempty [CondBranch c pcs e])
+                <$> parseConditionConfVar args
+                <*> subpcs
+                <*> elseClauses
+        pure (Just <$> condNode, rest)
+      _ -> (\r -> (pure Nothing, r)) <$> go [] x
+
+    parseImport :: Position -> [FieldLine Position] -> ParseResult FilePath
+    parseImport pos lines' = runFieldParser pos (P.many P.anyChar) cabalSpec lines'
+
+    -- We want a normalized path for @fieldsToConfig@. This eventually surfaces
+    -- in solver rejection messages and build messages "this build was affected
+    -- by the following (project) config files" so we want all paths shown there
+    -- to be relative to the directory of the project, not relative to the file
+    -- they were imported from.
+    fieldsToConfig :: ProjectConfigPath -> [Field Position] -> ParseResult ProjectConfig
+    fieldsToConfig sourceConfigPath xs = do
+      let (fs, sectionGroups) = partitionFields xs
+          sections = concat sectionGroups
+      config <- parseFieldGrammar cabalSpec fs (projectConfigFieldGrammar sourceConfigPath (knownProgramNames programDb))
+      config' <- view stateConfig <$> execStateT (goSections programDb sections) (SectionS config)
+      return config'
+
+    fetchImportConfig :: ProjectConfigPath -> IO BS.ByteString
+    fetchImportConfig (ProjectConfigPath (pci :| _)) = do
+      debug verbosity $ "fetching import: " ++ pci
+      fetch pci
+
+    fetch :: FilePath -> IO BS.ByteString
+    fetch pci = case parseURI pci of
```

Comment:
This is missing the trim of #10629.

https://github.com/haskell/cabal/blob/45e3289bce8e057d83421ed62bcae15ada290269/cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs#L350

Reply:
https://github.com/haskell/cabal/pull/10629 is merged now. @jgotoh any thoughts here?